### PR TITLE
Markups for PR 11

### PR DESCRIPTION
**Pull request by Brondahl**
Involved history modification, so done as a separate PR.

Personally don't really think there's anything in here that needs reviewing, but putting it up in case you disagree

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -187,46 +189,63 @@ Please examine the XML docs for full details. However, in simplified form, we ha
...
+                    .WithSequentialFrom(m => m.Colour, "Green", "Red", "Blue")                          // Monster Colors will be Green, Red, Blue, Green, Red, ...
 
-                    .WithCreateEnumerableFrom(m => m.Sounds, "Rarrrgggh!", "Screech!", "Wooooosh!") // All monsters will produce ALL three of these sounds.
-                    // Above is identical to ".WithCreateEnumerableFrom(m => m.Sounds, new List<string>{"Rarrrgggh!", "Screech!", "Woooooh!"})"
-                    // Above is *almost* identical to ".With(m => m.Sounds, new []{"Rarrrgggh!", "Screech!", "Woooooh!"})". The only difference is that the containing array is not shared.
+                    .WithCreateEnumerableFrom(m => m.Sounds, "Rarrrgggh!", "Screech!", "Wooooosh!")     // All monsters will produce all three of these sounds.
+                                                                                                        // Above is identical to ".WithCreateEnumerableFrom(m => m.Sounds, new List<string>{"Rarrrgggh!", "Screech!", "Woooooh!"})"
```
Comment (dhedey) [inline README.md]: This comment is self-referential, I'd probably prefer:
```
// This is equivalent to ".WithCreateEnumerableFrom(m => m.Sounds, new List<string>{"Rarrrgggh!", "Screech!", "Woooooh!"})"
// And to .WithFactory(m => m.Sounds, () => new List<string>{"Rarrrgggh!", "Screech!", "Wooooosh!"})
```

To be honest, I'm not sure I like the `WithCreateEnumerableFrom` method at all - I'd much rather have someone use `WithFactory` which is more obviously clear in what it does. I guess `WithCreateEnumerableFrom` can also be used to make shallow clones of the provided enumerable, so they don't share the same reference, but this feels a little unexpected - instead, they could simply use `.ToList()` to make the shallow clone (?)

File: README.md [hunk 2]
```diff
@@ -187,46 +189,63 @@ Please examine the XML docs for full details. However, in simplified form, we ha
...
+                                                                                                        // and *almost* identical to ".With(m => m.Sounds, new []{"Rarrrgggh!", "Screech!", "Woooooh!"})" (only difference is that the containing array is not shared.)
```
Comment (dhedey) [inline README.md]: I don't think this compiles any more - this would need to be `WithSharedRef`

File: README.md [hunk 3]
```diff
@@ -187,46 +189,63 @@ Please examine the XML docs for full details. However, in simplified form, we ha
...
+                    .WithFactory(m => m.FavouriteFood, () => new List<string>())                        // All monsters will get their own, distinct (initially empty) List<> object for food.
+                    
+                    .WithAddToCollection(m => m.FavouriteFood, "People")                                // All monsters like to eat people, in addition to anything that could have been configured prior to this point.
+                                                                                                        // (Obviously the more natural way to achieve that would be to include it in the previous Factory, but we want to demonstrate this .WithAddToCollection method.)
+
+                    .WithFactory(m => m.Age, () => rand.Next(6))                                        // Age might be 2, 4, 1, 4, 6, 3 ...
+                    
+                    .WithBuilder(m => m.HomeLake, LakeBuilder.Default)                                      // All monsters will have this.HomeLake populated with the result of "LakeBuilder.New.Build()"
```
Comment (dhedey) [inline README.md]: Again, I'm not sure if this method (or `WithNew` actually provide anything over `WithFactory` tbh, but maybe that's just me being a simplicist :)

File: README.md [hunk 4]
```diff
@@ -145,25 +146,25 @@ Please examine the XML docs for full details. However, in simplified form, we ha
   * Provide a factory method, to set a property to a newly created value each time.
 * ##### `WithBuilder()`
   * Like `WithFactory`, but the factory is specifically the `.Build()` method of the provided `Builder`.
-* ##### `WithBuilt()`
-  * Like `WithBuilder`, but the appropriate `Builder<T>` is found automatically from any classes tagged with `BuilderFactory`, so you don't need to provide any second argument at all ... it just works it all out.
 * ##### _`IEnumerable` variations of `WithFactory()`_
```
Comment (dhedey) [inline README.md:155]: Do these still exist? Or are they subsumed by `WithCreateEnumerableFrom`?

File: README.md (hunk 3 above)
Comment (dhedey) [reply README.md]: I guess if we were to remove things like `WithBuilder` and `WithNew`, we could just deprecate them for now; and even if we were to remove them from the core API, users could they could easily back-fill them as extension methods if they wanted them.

File: README.md [hunk 5]
```diff
@@ -145,25 +146,25 @@ Please examine the XML docs for full details. However, in simplified form, we ha
   * Provide a factory method, to set a property to a newly created value each time.
```
Comment (dhedey) [inline README.md]: I'd be tempted to have `WithFactory` up under `WithSharedRef`, as when I used it, I used `WithFactory` all the time, as it is the most powerful/flexible/general method

File: README.md [hunk 6]
```diff
@@ -312,10 +349,6 @@ public void WeirdStatefulBuilder()
 
 Note that such usages are stretching this library beyond the bounds of what it is intended for. You're likely better off building all the monsters up-front, and then updating them after the fact!
 
-* `With(m => m.SubObject, someExistingObject)` intended re-use the same object on every TInstance, is now `WithSharedRef(m => m.SubObject, someExistingObject)`.
-* `Add(...)` is now `WithAddToCollection(...)`.
-
-The broad changes to the API were to avoid overload conflicts, to improve clarity, and to try to improve discoverability of the available options.
 ## TODOs
```
Comment (dhedey) [inline README.md:359]: Should this be under a section called # Contributing or something?

File: README.md [hunk 7]
```diff
@@ -285,6 +303,25 @@ Please examine the XML docs for full details. However, in simplified form, we ha
...
+* You can use the `.WithBuilder()` methods to setup complex sub-properties, for which you've already defined `Builder`s.
+* In general, the various setup actions are executed against each object being built in the order in which they are configured on the Builder, and ALL actions are performed, even if they would be overridden by later setup actions.
+  * Thus if you configure `.With(m => m.Prop, val)` multiple times on the builder, then the resultant object will have the last value that was configured, but would have triggered any setter code associated with `m.Prop` repeatedly.
+* If you have a database object with properties representing DB relations, where there is a FK int property AND a FK object property (and possibly also a collection property on the other end of the relationship), then you probably want to use `.WithPostBuildSetup()` to ensure that everything gets suitably wired up at the end, to account for later modifications applied to the Builder.
+* `Builder` objects are immutable; each configuration method call derives a *new* `Builder` object, leaving the original one unchanged.
```
Comment (dhedey) [inline README.md:328]: I think this is really important and should be moved up to the introduction

Comment (dhedey) [reply README.md:328]: (And maybe even repeated under `Setup Methods`)

**Review by dhedey — APPROVED** [review]
Looks good! Have added a few more opinionated thoughts in case they resonate with you, but most of them don't need addressing! (not least as they're so late in the day).

Also apologies for the first half of my comments not being in the context of this review!

File: README.md (hunk 3 above)
Comment (Brondahl) [reply README.md]: I disagree.
This whole library is about making it easier to do these things. Adding helper short-hands for common versions of things that **_could_** be done other ways seems entirely in-scope.

(Plus, if an individual user doesn't think they're clear at the point of usage they can just ignore them)

Comment (dhedey) [reply README.md]: The comment should read: `LakeBuilder.Default.Build()`

Comment (Brondahl) [reply README.md]: 👍 Fixed.

File: README.md (hunk 2 above)
Comment (Brondahl) [reply README.md]: Great catch! ~Fixed~ Replaced by `WithFactory`, per your comment above.

File: README.md (hunk 1 above)
Comment (Brondahl) [reply README.md]: Hmm ... one of the main things it's doing is getting rid of irrelevant boilerplate of the details of the IEnumerable.
When you're setting up the test object you don't need to care about whether it's a List, Array or a ReadOnlyCollection - the library handles all that for you. But the docs here don't do a good job of making that clear.

I'll expand the comments to highlight that.

And if the user thinks it's not clear enough, they can just use the core method :)

File: README.md (hunk 4 above)
Comment (Brondahl) [reply README.md:155]: Yes, this exists and is distinct from `WithCreateEnumerableFrom`.

File: README.md (hunk 5 above)
Comment (Brondahl) [reply README.md]: 👍  Done.

File: README.md (hunk 6 above)
Comment (Brondahl) [reply README.md:359]: Do you mean the TODO?
Given that it's a single line, and I have plans to do it once we've got v4 out of the door, It doesn't seem worth changing

File: README.md (hunk 7 above)
Comment (Brondahl) [reply README.md:328]: 👍 Done.

File: README.md (hunk 1 above)
Comment (Brondahl) [reply README.md]: Done.

**Comment by Brondahl** [comment]
> 
> 
> Looks good! Have added a few more opinionated thoughts in case they resonate with you, but most of them don't need addressing! (not least as they're so late in the day).
> 
> Also apologies for the first half of my comments not being in the context of this review!

No problem :D

I've applied most of them, and disagreed with a handful (mostly "Do we really need this method?" ones. For which my answer is:
* I think they're useful.
* I think they're (sufficiently) clear, so as to not be actively detrimental to a codebase.
* They have no current cost to maintain.
* Users can ignore them if they like.
)

File: README.md (hunk 6 above)
Comment (dhedey) [reply README.md:359]: The heading "TODOs" is (I think) about contributing to the project and/or planned roadmap for the project, and not of interest to an average user deciding if they should use the module.

IMO these should be separate parts of the readme, or in a separate section (eg on the github wiki or issues/board?)  :)

But also, it's not a big deal 👍

Comment (Brondahl) [reply README.md:359]: Gotcha. Agreed, but won't delay 4.0 for it.
The 'How to publish' should likewise get moved.

**Comment by Brondahl** [comment]
All work reviewed, markups applied, code merged and latest version released.

**Pull request closed without merging**